### PR TITLE
Fix delete curve vertex

### PR DESCRIPTION
**Pull request by Djedouas**
## Description

Fixes a bug when deleting a middle curve vertex.

The deletion was not correctly handled, corrupting the geometry (see picture below, where using the vertex tool we can see a rubber band issue, persisting even after saving the layer), and creating an OGR error when exporting in GPKG (Non contiguous curves).

**Vertex to delete**

image

**After deletion**

image

## Fix

Test if the vertex to delete is a middle vertex of a curve, and if so, delete the curve and simply replace it with a linestring going from the first to the last vertex and skipping the middle one.

---

Funded by Métropole de Lille

### CODE REVIEW

File: src/core/geometry/qgscompoundcurve.cpp [hunk 1]
```diff
@@ -794,94 +794,131 @@ bool QgsCompoundCurve::moveVertex( QgsVertexId position, const QgsPoint &newPos
...
-         mCurves.at( curveIds.at( 1 ).first )->numPoints() > 3 )
+    int nextCurveId = curveIds.at( 1 ).first;
+    QgsCurve *nextCurve = mCurves.at( nextCurveId );
+    QgsVertexId nextSubVertexId = curveIds.at( 1 ).second;
+
+    Q_ASSERT( nextCurveId == curveId + 1 );
+    Q_ASSERT( subVertexId.vertex == curve->numPoints() - 1 );
+    Q_ASSERT( nextSubVertexId.vertex == 0 );
+
+    QgsPoint startPoint = curve->startPoint();
```
Comment (troopa81) [inline src/core/geometry/qgscompoundcurve.cpp:855]:
```suggestion
    const QgsPoint startPoint = curve->startPoint();
```

File: src/core/geometry/qgscompoundcurve.cpp [hunk 2]
```diff
@@ -794,94 +794,131 @@ bool QgsCompoundCurve::moveVertex( QgsVertexId position, const QgsPoint &newPos
...
+    QgsPoint endPoint = nextCurve->endPoint();
```
Comment (troopa81) [inline src/core/geometry/qgscompoundcurve.cpp:856]:
```suggestion
    const QgsPoint endPoint = nextCurve->endPoint();
```

**Review by troopa81 — CHANGES_REQUESTED** [review]
Does this PR fix also the following deleting issue ?

[errordel.webm](https://github.com/qgis/QGIS/assets/14358135/8553077c-eb43-41fb-91ec-d9ce61bf6eb1)

I didn't test it but looking at the code, I would say no. Am I right?

File: src/core/geometry/qgscompoundcurve.cpp [hunk 3]
```diff
@@ -794,94 +794,131 @@ bool QgsCompoundCurve::moveVertex( QgsVertexId position, const QgsPoint &newPos
...
+
+        std::unique_ptr<QgsCircularString> curveA = std::make_unique<QgsCircularString>();
+        curveA->setPoints( partA );
+        std::unique_ptr<QgsLineString> curveB = std::make_unique<QgsLineString>();
+        curveB->setPoints( partB );
+        std::unique_ptr<QgsCircularString> curveC = std::make_unique<QgsCircularString>();
+        curveC->setPoints( partC );
+
+        removeCurve( curveId );
+        if ( subVertexId.vertex < points.length() - 2 )
```
[the remark this answers is not in the document]
Comment (Djedouas) [reply src/core/geometry/qgscompoundcurve.cpp:818]: Yes you're right, I added a commit for this.

**Review by troopa81 — APPROVED** [review]

**Comment by troopa81** [comment]
There is still another issue to fix but current behavior is improved with this PR

**Pull request merged**